python/libff/kv.py

```diff
@@ -155,35 +155,67 @@ class Shmm(kv):
     """ A local-like kv store. With python shared_memory package.
     Not allowed to modify the existing values."""
 
-    def __init__(self, size=4096, serialize=True):
+    def __init__(self, serialize=True):
+        """ Currently, value must be binary."""
         """ Currently, value must be binary."""
         self.serialize = serialize
-        self.size = size
-        self.map = {}   # key: name; value: (offset, number of bytes)
-        self.shm = shared_memory.SharedMemory(create=True, size=size)
-        self.offset = 0
+        self.offset = 10
+        #self.map = {} or, we can serialize it and put it in the shmm
+        # key: name; value: (offset, number of bytes)
+        # In context manager, start posix_ipc and terminate
+        # [10 offset, 10 key, 10 size, val, 10 key, 10 size, val, ...]
+        self.shm = posix_ipc.SharedMemory("share")
+        self.mm = mmap.mmap(self.shm.fd, self.shm.size)
+        self.shm.close_fd()
+        self.sema = posix_ipc.Semaphore("share")
 
     def put(self, k, v, profile=None, profFinal=True):
+        #if k in self.map.keys():
+        #    raise ValueError("Duplicate key")
+        if len(k) > 10:
+            raise ValueError("length of key exceeds")
         with timer("t_serialize", profile, final=profFinal):
             if self.serialize:
                 v = pickle.dumps(v)
         num_bytes = len(v)
-        if self.offset + num_bytes >= self.size:
+        self.sema.acquire()
+        self.offset = int(self.mm[:10].rstrip(b'\x00'))
+        total = self.offset + num_bytes + 20    # careful if total exceeds 10 bytes
+        self.mm.seek(0)
```

Comment:
I think you can replace all the seeks with indexing (like you do in get()). I can't be sure, but I'd suspect that seek has some overheads associated with it. It's probably best not to have any seeks.